### DISCUSSION TITLE
require.extensions is not supported by webpack. Use a loader instead.

### DISCUSSION THREAD

**Discussion by arnelamo** [vercel/next.js#52593] [Help]
### Summary

I use a library called `nodemailer-express-handlebars` to generate HTML templates, and all though everything seems to work just fine - the HTML templates is generated as expected - i get this warning in the terminal when I build, and when I call the api endpoint in dev mode: 

`Import trace for requested module:
./node_modules/handlebars/lib/index.js
./node_modules/express-handlebars/dist/express-handlebars.js
./node_modules/express-handlebars/dist/index.js
./node_modules/nodemailer-express-handlebars/lib/generator.js
./node_modules/nodemailer-express-handlebars/lib/index.js
./node_modules/nodemailer-express-handlebars/index.js
./app/api/contactForm/route.ts

./node_modules/handlebars/lib/index.js
require.extensions is not supported by webpack. Use a loader instead.`

This image shows how I import the package in my `app/api/contactForm/route.ts` file:
<img width="675" alt="Screenshot 2023-07-11 at 18 45 48" src="https://github.com/vercel/next.js/assets/48482148/f119cea2-b262-49d2-8140-70965d6ceb2b">

This image shows my package.json:
<img width="488" alt="Screenshot 2023-07-12 at 15 33 09" src="https://github.com/vercel/next.js/assets/48482148/60daa13d-739f-47de-a8a1-7036e5a8d5cd">

I've tried to look for solutions, and found things like this:
[](https://github.com/handlebars-lang/handlebars.js/issues/1174
)

However, I'm not sure this is possible and even the right way of doing it. I had to add the `express-handlebars` to make the this work at all, and everything seems to work just fine. The warning however is bugging me, and I would like to know if this is possible to solve. I tried to look into the [docs](https://nextjs.org/docs/app/api-reference/next-config-js/webpack) but still can't figure out if I can or should try to configure an 'alias' somehow in the `next.config.js`.

All input is highly appreciated.

**Comment by icyJoseph** [comment]
I think this is the kind of thing that happens when you have several layers of abstraction over a package, not saying it is inherently bad, but the problem seems to be that webpack is trying to compile the underlying handlebars code.

- https://github.com/handlebars-lang/handlebars.js/issues/953#issuecomment-94931306

This comment has a webpack object you could try to replicate

- https://github.com/handlebars-lang/handlebars.js/issues/953#issuecomment-397228086

The last link is the same but updated, I think you can just paste that int the next.js config webpack key.

**Reply by arnelamo** [reply]
Awesome, I will try that. 

I did however try to change my next.config.js like this, just by copying the last 'import trace' from the log. You can see the changes I've made highlighted in green here.

<img width="581" alt="Screenshot 2023-07-12 at 16 12 16" src="https://github.com/vercel/next.js/assets/48482148/67e20ace-2a53-46fd-8bbb-88d3b3ae46f4">

Now, the warning is _actually_ gone. Would you by any chance like to comment on my "home made" solution?

**Reply by icyJoseph** [reply]
Just switch `var path` for `const path` hehe, but otherwise, it looks alright. 

Now sure though if telling Wepback to load `express-handlebars` whenever it finds `require("handlebars")` is a good idea 😅 it might just work now, but if a feature actually requires handlebars code, it'll most likely break.

The "problem" is that require loads the entire module synchronously, and with it, all of its required code, and so on... even if the code is unused.

**Comment by softmarshmallow** [comment]
what about for turbopack?

below wont' work

```ts
  turbopack: {
    resolveAlias: {
      handlebars: "handlebars/dist/handlebars.js",
    },
  },
```

**Reply by monokaijs** [reply]
Try add this to next.config.js:
```
  webpack: (config, context) => {
    return {
      ...config,
      resolve: {
        ...config.resolve,
        alias: {
          ...config.resolve.alias,
          handlebars: "handlebars/dist/handlebars.js",
        }
      }
    };
  },
  ```
